### ISSUE TITLE
Replica count

### ISSUE THREAD

**Issue by zheliazkov** [OpenNebula/addon-linstor#11]
Hello,

I'm experiencing strange behavior creating VMs, that did not had before (while using Ubuntu OS with deb installers and thin LVM):

The system datastore with ID 101 in OpenNebula is configured as folllows:
```
ALLOW_ORPHANS	yes	 
BRIDGE_LIST	colo-srv-1 colo-srv-2 colo-srv-3	 
DISK_TYPE	FILE	 
DS_MIGRATE	YES	 
LINSTOR_CLONE_MODE	copy	 
LINSTOR_CONTROLLERS	colo-srv-1,colo-srv-2,colo-srv-3	 
LINSTOR_RESOURCE_GROUP	OneSysRscGrp	 
RESTRICTED_DIRS	/	 
SAFE_DIRS	/var/tmp	 
SHARED	YES	 
TM_MAD	linstor	 
TYPE	SYSTEM_DS	 
```

The image datastore with ID 102 is configured there as follows:
```
ALLOW_ORPHANS	yes	 
BRIDGE_LIST	colo-srv-1 colo-srv-2 colo-srv-3	 
CLONE_TARGET	SELF	 
CLONE_TARGET_SHARED	SELF	 
CLONE_TARGET_SSH	SELF	 
COMPATIBLE_SYS_DS	0,101	 
DISK_TYPE	BLOCK	 
DISK_TYPE_SHARED	BLOCK	 
DISK_TYPE_SSH	BLOCK	 
DS_MAD	linstor	 
LINSTOR_CLONE_MODE	copy	 
LINSTOR_CONTROLLERS	colo-srv-1,colo-srv-2,colo-srv-3	 
LINSTOR_RESOURCE_GROUP	OneRscGrp	 
LN_TARGET	NONE	 
LN_TARGET_SHARED	NONE	 
LN_TARGET_SSH	NONE	 
RESTRICTED_DIRS	/	 
SAFE_DIRS	/var/tmp	 
TM_MAD	linstor	 
TM_MAD_SYSTEM	ssh,shared	 
TYPE	IMAGE_DS	 
```

I've decided to use non-thin LVM, so I set **LINSTOR_CLONE_MODE** to **copy**

The resource groups in the Linstor are created as follows:
```
╭────────────────────────────────────────────────────────────────╮
┊ ResourceGroup ┊ SelectFilter            ┊ VlmNrs ┊ Description ┊
╞════════════════════════════════════════════════════════════════╡
┊ DfltRscGrp    ┊ PlaceCount: 2           ┊        ┊             ┊
╞┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄╡
┊ OneRscGrp     ┊ PlaceCount: 3           ┊ 0      ┊             ┊
┊               ┊ StoragePool(s): onepool ┊        ┊             ┊
╞┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄╡
┊ OneSysRscGrp  ┊ PlaceCount: 3           ┊ 0      ┊             ┊
┊               ┊ StoragePool(s): onepool ┊        ┊             ┊

```

The question is why the images are placed on 3 storage nodes, but when I create VMs their disks/resources are placed only on two of them?

Maybe there is some mistake I do but cannot find a reason.

Thanks in advance.

BR,
Plamen

**Comment by rp-** [comment]
do the VM resources belong to the correct resource groups? should be shown by `linstor rd l`

**Comment by zheliazkov** [comment]
> do the VM resources belong to the correct resource groups? should be shown by `linstor rd l`

I see that when deploying VM its context disk is on the System datastore but the disk is on the other one(Image datastore).
Both resource groups have place count 3.

```
[root@colo-srv-2 ~]# linstor --controller colo-srv-3 rd l
╭───────────────────────────────────────────────────────────────────╮
┊ ResourceName                       ┊ Port ┊ ResourceGroup ┊ State ┊
╞═══════════════════════════════════════════════════════════════════╡
┊ OpenNebula-Image-14                ┊ 7004 ┊ OneRscGrp     ┊ ok    ┊
┊ OpenNebula-Image-14-vm46-disk0     ┊ 7009 ┊ OneRscGrp     ┊ ok    ┊
┊ OpenNebula-Image-14-vm47-disk0     ┊ 7005 ┊ OneRscGrp     ┊ ok    ┊
┊ OpenNebula-Image-14-vm50-disk0     ┊ 7007 ┊ OneRscGrp     ┊ ok    ┊
┊ OpenNebula-Image-14-vm51-disk0     ┊ 7010 ┊ OneRscGrp     ┊ ok    ┊
┊ OpenNebula-Image-14-vm52-disk0     ┊ 7011 ┊ OneRscGrp     ┊ ok    ┊
┊ OpenNebula-Image-14-vm56-disk0     ┊ 7012 ┊ OneRscGrp     ┊ ok    ┊
┊ OpenNebula-Image-14-vm57-disk0     ┊ 7013 ┊ OneRscGrp     ┊ ok    ┊
┊ OpenNebula-Image-14-vm58-disk0     ┊ 7015 ┊ OneRscGrp     ┊ ok    ┊
┊ OpenNebula-Image-14-vm59-disk0     ┊ 7017 ┊ OneRscGrp     ┊ ok    ┊
┊ OpenNebula-Image-6                 ┊ 7001 ┊ OneRscGrp     ┊ ok    ┊
┊ OpenNebula-Image-7                 ┊ 7003 ┊ OneRscGrp     ┊ ok    ┊
┊ OpenNebula-Image-8                 ┊ 7008 ┊ OneRscGrp     ┊ ok    ┊
┊ OpenNebula-Image-9                 ┊ 7006 ┊ OneRscGrp     ┊ ok    ┊
┊ OpenNebula-Image-9-vm44-disk0      ┊ 7002 ┊ OneRscGrp     ┊ ok    ┊
┊ OpenNebula-vm-context-vm57-disk1-1 ┊ 7014 ┊ OneSysRscGrp  ┊ ok    ┊
┊ OpenNebula-vm-context-vm58-disk1-1 ┊ 7016 ┊ OneSysRscGrp  ┊ ok    ┊
┊ OpenNebula-vm-context-vm59-disk1-1 ┊ 7018 ┊ OneSysRscGrp  ┊ ok    ┊
┊ linstor_db                         ┊ 7000 ┊ DfltRscGrp    ┊ ok    ┊
╰───────────────────────────────────────────────────────────────────╯

```
The resource placement is as follows:
```
[root@colo-srv-3 ~]# linstor --controller colo-srv-3 r l
╭──────────────────────────────────────────────────────────────────────────────────────────────────────────╮
┊ ResourceName                       ┊ Node       ┊ Port ┊ Usage  ┊ Conns ┊    State ┊ CreatedOn           ┊
╞══════════════════════════════════════════════════════════════════════════════════════════════════════════╡
┊ OpenNebula-Image-14                ┊ colo-srv-1 ┊ 7004 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-28 12:36:01 ┊
┊ OpenNebula-Image-14                ┊ colo-srv-2 ┊ 7004 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-28 12:36:01 ┊
┊ OpenNebula-Image-14                ┊ colo-srv-3 ┊ 7004 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-28 12:36:01 ┊
┊ OpenNebula-Image-14-vm46-disk0     ┊ colo-srv-1 ┊ 7009 ┊ InUse  ┊ Ok    ┊ UpToDate ┊ 2021-04-28 17:25:14 ┊
┊ OpenNebula-Image-14-vm46-disk0     ┊ colo-srv-2 ┊ 7009 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-28 17:25:16 ┊
┊ OpenNebula-Image-14-vm47-disk0     ┊ colo-srv-2 ┊ 7005 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-28 17:25:15 ┊
┊ OpenNebula-Image-14-vm47-disk0     ┊ colo-srv-3 ┊ 7005 ┊ InUse  ┊ Ok    ┊ UpToDate ┊ 2021-04-28 17:25:14 ┊
┊ OpenNebula-Image-14-vm50-disk0     ┊ colo-srv-1 ┊ 7007 ┊ InUse  ┊ Ok    ┊ UpToDate ┊ 2021-04-28 19:10:49 ┊
┊ OpenNebula-Image-14-vm50-disk0     ┊ colo-srv-2 ┊ 7007 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-28 19:10:50 ┊
┊ OpenNebula-Image-14-vm51-disk0     ┊ colo-srv-2 ┊ 7010 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-28 19:10:50 ┊
┊ OpenNebula-Image-14-vm51-disk0     ┊ colo-srv-3 ┊ 7010 ┊ InUse  ┊ Ok    ┊ UpToDate ┊ 2021-04-28 19:10:49 ┊
┊ OpenNebula-Image-14-vm52-disk0     ┊ colo-srv-1 ┊ 7011 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-28 19:23:54 ┊
┊ OpenNebula-Image-14-vm52-disk0     ┊ colo-srv-2 ┊ 7011 ┊ InUse  ┊ Ok    ┊ UpToDate ┊ 2021-04-28 19:23:53 ┊
┊ OpenNebula-Image-14-vm56-disk0     ┊ colo-srv-1 ┊ 7012 ┊ InUse  ┊ Ok    ┊ UpToDate ┊ 2021-04-29 10:33:57 ┊
┊ OpenNebula-Image-14-vm56-disk0     ┊ colo-srv-3 ┊ 7012 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-29 10:33:58 ┊
┊ OpenNebula-Image-14-vm57-disk0     ┊ colo-srv-1 ┊ 7013 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-29 11:08:41 ┊
┊ OpenNebula-Image-14-vm57-disk0     ┊ colo-srv-3 ┊ 7013 ┊ InUse  ┊ Ok    ┊ UpToDate ┊ 2021-04-29 11:08:40 ┊
┊ OpenNebula-Image-14-vm58-disk0     ┊ colo-srv-2 ┊ 7015 ┊ InUse  ┊ Ok    ┊ UpToDate ┊ 2021-05-05 13:11:29 ┊
┊ OpenNebula-Image-14-vm58-disk0     ┊ colo-srv-3 ┊ 7015 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-05-05 13:11:30 ┊
┊ OpenNebula-Image-6                 ┊ colo-srv-1 ┊ 7001 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-26 11:14:35 ┊
┊ OpenNebula-Image-6                 ┊ colo-srv-2 ┊ 7001 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-26 11:14:35 ┊
┊ OpenNebula-Image-6                 ┊ colo-srv-3 ┊ 7001 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-26 11:14:35 ┊
┊ OpenNebula-Image-7                 ┊ colo-srv-1 ┊ 7003 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-26 11:28:17 ┊
┊ OpenNebula-Image-7                 ┊ colo-srv-2 ┊ 7003 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-26 11:28:17 ┊
┊ OpenNebula-Image-7                 ┊ colo-srv-3 ┊ 7003 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-26 11:28:17 ┊
┊ OpenNebula-Image-8                 ┊ colo-srv-1 ┊ 7008 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-26 12:09:22 ┊
┊ OpenNebula-Image-8                 ┊ colo-srv-2 ┊ 7008 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-26 12:09:22 ┊
┊ OpenNebula-Image-8                 ┊ colo-srv-3 ┊ 7008 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-26 12:09:22 ┊
┊ OpenNebula-Image-9                 ┊ colo-srv-1 ┊ 7006 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-27 17:02:59 ┊
┊ OpenNebula-Image-9                 ┊ colo-srv-2 ┊ 7006 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-27 17:02:59 ┊
┊ OpenNebula-Image-9                 ┊ colo-srv-3 ┊ 7006 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-27 17:02:59 ┊
┊ OpenNebula-Image-9-vm44-disk0      ┊ colo-srv-1 ┊ 7002 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-28 17:22:46 ┊
┊ OpenNebula-Image-9-vm44-disk0      ┊ colo-srv-2 ┊ 7002 ┊ InUse  ┊ Ok    ┊ Diskless ┊ 2021-04-28 17:22:45 ┊
┊ OpenNebula-Image-9-vm44-disk0      ┊ colo-srv-3 ┊ 7002 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-28 17:22:44 ┊
┊ OpenNebula-vm-context-vm57-disk1-1 ┊ colo-srv-1 ┊ 7014 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-29 11:08:53 ┊
┊ OpenNebula-vm-context-vm57-disk1-1 ┊ colo-srv-2 ┊ 7014 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-29 11:08:53 ┊
┊ OpenNebula-vm-context-vm57-disk1-1 ┊ colo-srv-3 ┊ 7014 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-29 11:08:53 ┊
┊ OpenNebula-vm-context-vm58-disk1-1 ┊ colo-srv-1 ┊ 7016 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-05-05 13:11:42 ┊
┊ OpenNebula-vm-context-vm58-disk1-1 ┊ colo-srv-2 ┊ 7016 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-05-05 13:11:42 ┊
┊ OpenNebula-vm-context-vm58-disk1-1 ┊ colo-srv-3 ┊ 7016 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-05-05 13:11:42 ┊
┊ linstor_db                         ┊ colo-srv-1 ┊ 7000 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-21 13:41:10 ┊
┊ linstor_db                         ┊ colo-srv-2 ┊ 7000 ┊ Unused ┊ Ok    ┊ UpToDate ┊ 2021-04-21 13:41:10 ┊
┊ linstor_db                         ┊ colo-srv-3 ┊ 7000 ┊ InUse  ┊ Ok    ┊ UpToDate ┊ 2021-04-21 13:41:12 ┊
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```
Thanks in advance.

BR,
Plamen

**Comment by zheliazkov** [comment]
@rp- 

I've been investigation the issue and found the following thing:
- If I download Image from the OpenNebula marketplace to the Image datastore (RG OneRscGroup) it has 3 replicas.
- If I make an Image from running VM it is placed only on two of them like the VM's images that are created on instancing the VMs.

But still don't know why...

BR,
Plamen

**Comment by rp-** [comment]
Sorry I didn't have yet time to look closer into the issue, but could you maybe provide a journal log `-u opennebula` from such a VM placement?

**Comment by zheliazkov** [comment]
@rp- 
> Sorry I didn't have yet time to look closer into the issue, but could you maybe provide a journal log `-u opennebula` from such a VM placement?

Thanks, here is the journal log of the opennebula in a placement with only two nodes:
```
15:56:55 colo-srv-1 /clone[76683]: INFO Entering tm clone src:colo-srv-1:OpenNebula-Image-24 dst:colo-srv-1:/var/lib/one//datastores/101/124/disk.0
15:56:55 colo-srv-1 /clone[76683]: INFO running shell command: onedatastore show --xml 102
15:56:56 colo-srv-1 /clone[76683]: INFO running shell command: onevm show -x 124
15:56:56 colo-srv-1 /clone[76683]: INFO Cloning from resource 'OpenNebula-Image-24' to 'OpenNebula-Image-24-vm124-disk0' clone mode copy.
15:56:56 colo-srv-1 /clone[76683]: INFO Deploying resource 'OpenNebula-Image-24-vm124-disk0' using resource group 'OneRscGrp', prefer node: colo-srv-1
15:56:58 colo-srv-1 /clone[76683]: INFO ssh 'colo-srv-1' cmd: dd if=/dev/drbd1004 of=/dev/drbd1001 bs=64K count=363712 conv=fsync
15:58:08 colo-srv-1 /clone[76683]: INFO ssh 'colo-srv-1' cmd: mkdir -p /var/lib/one/datastores/101/124 && ln -fs /dev/drbd1001 /var/lib/one/datastores/101/124/disk.0
15:58:09 colo-srv-1 /clone[76683]: INFO running shell command: oneimage show --xml 24
15:58:10 colo-srv-1 /clone[76683]: INFO ssh 'colo-srv-1' cmd: $QEMU_IMG resize /dev/drbd1001 102400M
15:58:11 colo-srv-1 /clone[76683]: INFO Exiting tm clone
15:58:11 colo-srv-1 /clone[77714]: INFO Entering tm clone src:colo-srv-1:OpenNebula-Image-25 dst:colo-srv-1:/var/lib/one//datastores/101/124/disk.1
15:58:11 colo-srv-1 /clone[77714]: INFO running shell command: onedatastore show --xml 102
15:58:11 colo-srv-1 /clone[77714]: INFO running shell command: onevm show -x 124
15:58:12 colo-srv-1 /clone[77714]: INFO Cloning from resource 'OpenNebula-Image-25' to 'OpenNebula-Image-25-vm124-disk1' clone mode copy.
15:58:12 colo-srv-1 /clone[77714]: INFO Deploying resource 'OpenNebula-Image-25-vm124-disk1' using resource group 'OneRscGrp', prefer node: colo-srv-1
15:58:14 colo-srv-1 /clone[77714]: INFO ssh 'colo-srv-1' cmd: dd if=/dev/drbd1005 of=/dev/drbd1002 bs=64K count=65536 conv=fsync
15:58:29 colo-srv-1 /clone[77714]: INFO ssh 'colo-srv-1' cmd: mkdir -p /var/lib/one/datastores/101/124 && ln -fs /dev/drbd1002 /var/lib/one/datastores/101/124/disk.1
15:58:30 colo-srv-1 /clone[77714]: INFO Exiting tm clone
15:58:31 colo-srv-1 /context[78279]: INFO running shell command: onedatastore show --xml 101
15:58:31 colo-srv-1 /context[78279]: INFO Entering context on colo-srv-1:/var/lib/one//datastores/101/124/disk.2 for files ['/var/lib/one/vms/124/context.sh'].
15:58:31 colo-srv-1 /context[78279]: INFO copy /var/lib/one/vms/124/context.sh /var/lib/one//datastores/101/.isofiles/124/context.sh
15:58:31 colo-srv-1 /context[78279]: INFO running shell command: bash -c source /var/lib/one/remotes//scripts_common.sh && $MKISOFS -o /var/lib/one//datastores/101/.isofiles/124/124.iso -V CONTEXT -J -R //one//datastores/101/.isofiles/124
15:58:31 colo-srv-1 /context[78279]: INFO Deploying resource 'OpenNebula-vm-context-vm124-disk2-1' using resource group 'OneSysRscGrp', prefer node: None
15:58:33 colo-srv-1 /context[78279]: INFO running shell command: bash -c cat /var/lib/one//datastores/101/.isofiles/124/124.iso | ssh colo-srv-1 dd of=/dev/drbd1003 status=none
15:58:34 colo-srv-1 /context[78279]: INFO ssh 'colo-srv-1' cmd: mkdir -p /var/lib/one/datastores/101/124 && ln -fs /dev/drbd1003 /var/lib/one/datastores/101/124/disk.2
15:58:35 colo-srv-1 /context[78279]: INFO Leaving context successfully.
```

And this is when downloading new image from the marketplace when its placed on all nodes (3):
```
16:05:51 colo-srv-1 /cp[82229]: INFO Entering datastore cp
16:05:51 colo-srv-1 /cp[82229]: INFO Deploying resource 'OpenNebula-Image-27' using resource group 'OneRscGrp', prefer node: None
16:05:52 colo-srv-1 /cp[82229]: INFO running shell command: bash -c source /var/lib/one/remotes//datastore/libfs.sh && set_up_datastore /var/lib/one//datastores/102 / /var/tmp
16:05:52 colo-srv-1 /cp[82229]: INFO running shell command: bash -c source /var/lib/one/remotes//datastore/libfs.sh && set_downloader_args "" "" "" "" https://marketplace.opennebula.io/ce/8bf38934-1116-11ea-88e0-f0def1753696/download/0 -
16:05:52 colo-srv-1 /cp[82229]: INFO Downloading https://marketplace.opennebula.io/appliance/8bf38934-1116-11ea-88e0-f0def1753696/download/0 to the image repository
16:05:52 colo-srv-1 /cp[82229]: INFO running shell command: bash -c source /var/lib/one/remotes//datastore/libfs.sh && check_restricted https://marketplace.opennebula.io/ce/8bf38934-1116-11ea-88e0-f0def1753696/download/0
16:05:52 colo-srv-1 /cp[82229]: INFO Copying local image https://marketplace.opennebula.io/appliance/8bf38934-1116-11ea-88e0-f0def1753696/download/0 to the image repository
16:05:52 colo-srv-1 /cp[82229]: INFO running shell command: bash -c /var/lib/one/remotes//datastore/downloader.sh   'https://marketplace.opennebula.io/ce/8bf38934-1116-11ea-88e0-f0def1753696/download/0' '-' | ssh colo-srv-1 dd of=/dev/drbd1007 bs=2M oflag=direct
```
BR,
Plamen

**Comment by rp-** [comment]
Thanks again for reporting, it should be fixed with this commit 71cc6d5d625929f0350cec866ff07e953fcebe12

Could you test it?

**Comment by zheliazkov** [comment]
> Thanks again for reporting, it should be fixed with this commit [71cc6d5](https://github.com/OpenNebula/addon-linstor/commit/71cc6d5d625929f0350cec866ff07e953fcebe12)
> 
> Could you test it?

Thanks :) I'll test it this week and write back.

**Comment by zheliazkov** [comment]
@rp- I confirm that the commit fixed the problem.

Thanks for the help and your time.

BR,
Plamen
